### PR TITLE
Align ANN2SNN recipes with step-mode execution

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary

This PR aligns built-in ANN2SNN conversion products with SpikingJelly step-mode execution and adds layer-wise multi-step inference support across the built-in recipes.

Converted models are now ordinary `fx.GraphModule` products whose internal modules carry step-mode behavior. Users switch execution with `functional.set_step_mode(model, "s" | "m")`:

- `step_mode="s"`: user writes the outer time loop and calls the model one timestep at a time.
- `step_mode="m"`: user passes a sequence tensor whose first dimension is time and receives a sequence output.
- Readout is explicit: users sum the time dimension or apply their own structured-output readout.

This covers:

- `STATransformerRecipe`
- `TransformerSpikeEquivalentRecipe`
- `RateCodingRecipe`
- `LocalThresholdBalancingRecipe`

## Main Changes

- Refactor TD/SNN sequence operators to support stateful single-step semantics plus vectorized multi-step semantics.
- Replace STA execution with TD modules and a unified step-mode spike encoder where possible.
- Keep converted top-level products as ordinary graph modules; step-mode behavior lives in minimal internal modules.
- Add a shared step-mode graph adapter for built-in recipes.
- Wrap stateless CNN modules with existing `spikingjelly.activation_based.layer` step-mode modules for rate-coding/LTB paths.
- Add strict unsupported-op failures for unsafe FX tensor ops.
- Support ViT-style sequence-preserving tensor ops: shape/dim, batch-preserving reshape/permute, time-distributed expand, non-batch cat, and batch-preserving getitem.
- Preserve Transformer-path `nn.Dropout` instead of wrapping it as `layer.Dropout`.
- Update STA ViT ImageNet example to use explicit multi-step sequence input and explicit time readout.
- Add STA and RateCoding step-mode benchmark entry points.
- Update docs for explicit step-mode execution.

## Breaking / Migration Notes

This is a hard execution API change for STA converted products.

Old behavior hid the full `T` loop and implicit readout inside the converted model. New behavior follows SpikingJelly conventions:

```python
functional.set_step_mode(converted, "s")
functional.reset_net(converted)
y = None
for t in range(T):
    x_t = x if t == 0 else torch.zeros_like(x)
    y_t = converted(x_t)
    y = y_t if y is None else y + y_t
```

```python
functional.set_step_mode(converted, "m")
functional.reset_net(converted)
x_seq = torch.cat((x.unsqueeze(0), torch.zeros_like(x).expand(T - 1, *x.shape)))
y_seq = converted(x_seq)
y = y_seq.sum(dim=0)
```

For rate-coding/LTB, multi-step input is repeated static input `[x, x, ..., x]` rather than STA's `[x, 0, ..., 0]` differential input.

## Benchmark Notes

CUDA benchmark host: `g2`, `CUDA_VISIBLE_DEVICES=0`, NVIDIA A100-SXM4-80GB.

RateCoding benchmark:

- single-step loop: total `3.4099s`, epoch `0.6820s`, `6006.04` samples/s
- multi-step: total `3.0418s`, epoch `0.6084s`, `6732.85` samples/s
- speedup: `1.12x`

STA Transformer benchmark:

- single-step loop: total `15.6023s`, epoch `3.1205s`, `656.31` samples/s
- multi-step: total `13.9806s`, epoch `2.7961s`, `732.44` samples/s
- speedup: `1.12x`

A real ViT-B/16 ImageNet smoke was also run on `g2` A100 after the step-mode API change with `calib_samples=8`, `eval_samples=8`, `batch_size=4`, `time_steps=4`, `threshold_scale=0.5`; baseline and STA both reported top-1/top-5 `0.875 / 0.875` on that smoke subset.

## Validation

```bash
uv run python -m py_compile \
  spikingjelly/activation_based/ann2snn/modules.py \
  spikingjelly/activation_based/ann2snn/operators.py \
  spikingjelly/activation_based/ann2snn/recipes/rate_coding.py \
  spikingjelly/activation_based/ann2snn/recipes/local_threshold_balancing.py \
  spikingjelly/activation_based/ann2snn/recipes/transformer_spike_equivalent.py \
  spikingjelly/activation_based/ann2snn/recipes/sta_transformer.py \
  spikingjelly/activation_based/ann2snn/recipes/step_mode_adapters.py

uv run pytest \
  test/activation_based/test_ann2snn.py \
  test/activation_based/test_ann2snn_transformer.py \
  test/activation_based/test_ann2snn_operators.py -q

PYTHONPATH=$PWD uv run python -m sphinx -E -b html docs/source \
  /tmp/spikingjelly-docs-ann2snn-stepmode-check -W --keep-going

git diff --check
```

Latest local result: `437 passed, 1 warning` for the ANN2SNN pytest target.

## Summary by CodeRabbit

* **New Features**
  * Added standalone ANN→SNN benchmarking scripts for rate-coding and STA Transformer comparisons (single-step loop vs multi-step execution).
  * Added TDConv2d support for converted TD-based models.
* **Bug Fixes**
  * Improved delay/readout estimation edge-case handling and strengthened argument validation (including step-mode parity/reset expectations).
  * Enhanced evaluation/quantization behavior for voltage utilities.
* **Documentation**
  * Clarified step-mode execution semantics and reset requirements in tutorials and API notes.
* **Refactor**
  * Updated conversion/operator internals for consistent step-mode handling, including graph adaptation.
* **Tests**
  * Expanded ANN→SNN and operator coverage, including multi-step equivalence and TDConv2d validation.

### CODE REVIEW

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review
